File: NumericErrors/src/NumericErrors.java
public class NumericErrors {

    public static int add(int x, int y) throws ArithmeticException {
        // Check for positive overflow
        if (x > 0 && y > 0 && x > Integer.MAX_VALUE - y) {
            throw new ArithmeticException("Positive overflow detected in addition");
        }
        // Check for negative overflow
        if (x < 0 && y < 0 && x < Integer.MIN_VALUE - y) {
            throw new ArithmeticException("Negative overflow detected in addition");
        }
        return x + y;
    }

    public static int subtract(int x, int y) throws ArithmeticException {
        // Use add function to check for overflow
        return add(x, -y);
    }

    public static int multiply(int x, int y) throws ArithmeticException {
        if (y > 0 && x > Integer.MAX_VALUE / y) {
            throw new ArithmeticException("Positive overflow detected in multiplication");
        }
        if (y < 0 && x < Integer.MIN_VALUE / y) {
            throw new ArithmeticException("Negative overflow detected in multiplication");
        }
        return x * y;
    }

    public static int divide(int x, int y) throws ArithmeticException {
        if (y == 0) {
            throw new ArithmeticException("Division by zero");
        }
        if (x == Integer.MIN_VALUE && y == -1) {
            throw new ArithmeticException("Overflow detected in division");
        }
        return x / y;
    }

    public static void main(String[] args) {
        // Test cases can be added here
    }
}

File: exercise #2/C/NumericErrors.c
#include <stdio.h>
#include <limits.h>
#include <assert.h>
int add(int x, int y)
{
    // test
    // test
    //  Check for positive overflow
    if (x > 0 && y > 0 && x > INT_MAX - y)
    {
        printf("Positive overflow detected in addition!\n");
        exit(1);
    }
    // Check for negative overflow
    if (x < 0 && y < 0 && x < INT_MIN - y)
    {
        printf("Negative overflow detected in addition!\n");
        exit(1);
    }
    return x + y;
}

int subtract(int x, int y)
{
    // Use add function to check for overflow
    return add(x, -y);
}

int multiply(int x, int y)
{
    if (y > 0 && x > INT_MAX / y)
    {
        printf("Positive overflow detected in multiplication!\n");
        exit(1);
    }
    if (y < 0 && x < INT_MIN / y)
    {
        printf("Negative overflow detected in multiplication!\n");
        exit(1);
    }
    return x * y;
}

int divide(int x, int y)
{
    if (y == 0)
    {
        printf("Division by zero!\n");
        exit(1);
    }
    if (x == INT_MIN && y == -1)
    {
        printf("Overflow detected in division!\n");
        exit(1);
    }
    return x / y;
}

int main()
{
    // Test cases can be added here
    int a = 1;
    int b = 1;
    divide(a, b);
    add(a, b);
    return 0;
}
